### ISSUE TITLE
SPProductUpdate: ShutdownServices set to True not skipped it SPFarm not exists

### ISSUE THREAD

**Issue by paololec** [PowerShell/SharePointDsc#1087]
#### Details of the scenario you tried and the problem that is occurring
Executed SPProductUpdate before SPFarm with ShutdownServices  = $true produce errors.

#### Verbose logs showing the problem
[[SPProductUpdate]InstallCU] Check if BinaryInstallDays is specified
[[SPProductUpdate]InstallCU] No BinaryInstallDays specified, Update can be ran on any day.
[[SPProductUpdate]InstallCU] Check if BinaryInstallTime is specified
[[SPProductUpdate]InstallCU] No BinaryInstallTime specified, Update can be ran at any time. Starting update.
[[SPProductUpdate]InstallCU] Stopping services to speed up installation process
[[SPProductUpdate]InstallCU] Executing as the local run as user DEMO\SP_Admin
[[SPProductUpdate]InstallCU] Leaving BeginProcessing Method of Get-SPEnterpriseSearchServiceApplication.
SPFarm not found.
    + CategoryInfo          : InvalidData: (Microsoft.Offic...viceApplication:) [], CimException
    + FullyQualifiedErrorId : Microsoft.Office.Server.Search.Cmdlet.GetSearchServiceApplication
    + PSComputerName        : COMPUTERNAME
 
[[SPProductUpdate]InstallCU] Leaving ProcessRecord Method of Get-SPEnterpriseSearchServiceApplication.
[[SPProductUpdate]InstallCU] Leaving EndProcessing Method of Get-SPEnterpriseSearchServiceApplication.
[[SPProductUpdate]InstallCU] Search Services are stopped
[[SPProductUpdate]InstallCU] Stopping other services
[[SPProductUpdate]InstallCU] Beginning installation of the SharePoint update
[[SPProductUpdate]InstallCU] Executing as the local run as user DEMO\SP_Admin

[[SPProductUpdate]InstallCU] Restart stopped services
[[SPProductUpdate]InstallCU] Executing as the local run as user DEMO\SP_Admin
[[SPProductUpdate]InstallCU] Leaving BeginProcessing Method of Get-SPEnterpriseSearchServiceApplication.
SPFarm not found.
    + CategoryInfo          : InvalidData: (Microsoft.Offic...viceApplication:) [], CimException
    + FullyQualifiedErrorId : Microsoft.Office.Server.Search.Cmdlet.GetSearchServiceApplication
    + PSComputerName        : COMPUTERNAME
 
[[SPProductUpdate]InstallCU] Leaving ProcessRecord Method of Get-SPEnterpriseSearchServiceApplication.
[[SPProductUpdate]InstallCU] Leaving EndProcessing Method of Get-SPEnterpriseSearchServiceApplication.
[[SPProductUpdate]InstallCU] Services restarted.
LCM:  [ End    Set      ]  [[SPProductUpdate]InstallCU]  in 706.7750 seconds.
The PowerShell DSC resource '[SPProductUpdate]InstallCU' with SourceInfo 'C:\sourcepath\configure.ps1::1240::13::SPProductUpdate' threw one or more non-terminating 
errors while running the Set-TargetResource functionality. These errors are logged to the ETW channel called Microsoft-Windows-DSC/Operational. Refer to this channel for more details.
    + CategoryInfo          : InvalidOperation: (:) [], CimException
    + FullyQualifiedErrorId : NonTerminatingErrorFromProvider
    + PSComputerName        : COMPUTERNAME
 
#### Suggested solution to the issue
Ignore ShutdownServices  = $true if farm not exists.

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)
```powershell
SPProductUpdate InstallCU
            {
                SetupFile            =  pathtofile/stsxx.exe
                ShutdownServices     = $true
                Ensure               = "Present"
                PsDscRunAsCredential = $sp_admin
            }
BEFORE SPFarm
```

#### Version of SharePoint that is used (e.g. SharePoint 2016)
SharePoint 2019
#### Version and build of PowerShell the target node is running
PSVersion                      5.1.17763.503
#### Version of the DSC module that was used ('dev' if using current dev branch)
v3.5

**Comment by andikrueger** [comment]
Thanks for pointing this out. Will fix this issue.

**Comment by andikrueger** [comment]
Working on it.

**Comment by jonathaneckman** [comment]
This breaks SP 2013 builds. For some reason, the `ErrorAction` is ignored which means [this line](https://github.com/PowerShell/SharePointDsc/blob/release-3.6/Modules/SharePointDsc/DSCResources/MSFT_SPProductUpdate/MSFT_SPProductUpdate.psm1#L398):

```
$farm = Get-SPFarm -ErrorAction SilentlyContinue
```
fails every time. We had to fall back to 3.5 and set ShutdownServices to false. A try catch would work:

```
$result = "FARM"
try{
    $farm = Get-SPFarm -Verbose -ErrorAction SilentlyContinue
} catch {
    $result = "NO FARM"
}
Write-Host $result
```

**Comment by andikrueger** [comment]
Could you share your verbose log of a SPProductUpdate run?

`ErrorAction SilentlyContinue` should not break not throw an exception. @ykuijs is this a known error?

**Comment by ykuijs** [comment]
Some cmdlets do not handle the error action parameter correctly. In other resources we indeed use the try/catch method for Get-SPFarm:
https://github.com/PowerShell/SharePointDsc/blob/0c0ef48b972b3ab83cedbd5b2595270e33bd8a9a/Modules/SharePointDsc/DSCResources/MSFT_SPAntivirusSettings/MSFT_SPAntivirusSettings.psm1#L133-L141

**Comment by andikrueger** [comment]
I think we should add the try catch block to prevent this error.

**Comment by ykuijs** [comment]
Will implement this in the bugfix PR I am working on

**Comment by ykuijs** [comment]
This issue is now fixed in dev and will be released with v3.7
